### PR TITLE
Fix warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library

### LINKED ISSUE #729: Fix linker warning LNK4098: defaultlib 'LIBCMTD' conflicts ...
Similar to https://github.com/Azure/azure-sdk-for-c/pull/1394, fix the CRT version used to avoid the warning when building the SDK with tests.

From the build directory:
> cmake -DBUILD_TESTING=ON -DBUILD_CURL_TRANSPORT=ON ..

```text
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library [E:\GitHub\For
k\azure-sdk-for-cpp\build\sdk\storage\azure-storage-test.vcxproj]
  azure-storage-test.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\storage\Debug\azure-storage-test.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/template/azure-template/CMakeLists.txt
  version.cpp
  template_client.cpp
  Running Code Analysis for C/C++...
  Generating Code...
  azure-template.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\template\azure-template\Debug\azure-template.lib
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/template/azure-template/test/CMakeLists.txt
  template_test.cpp
  azure-template-test.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\template\azure-template\test\Debug\azure-te
  mplate-test.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/core/azure-core/test/e2e/CMakeLists.txt
  azure_core_storage_list_containers_sample.cpp
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library [E:\GitHub\For
k\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\azure_core_storage_issue_248.vcxproj]
  azure_core_storage_issue_248.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\Debug\azu
  re_core_storage_issue_248.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/core/azure-core/test/e2e/CMakeLists.txt
  azure_core_storage_test_sample.cpp
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library [E:\GitHub\For
k\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\azure_core_storage_issue_249.vcxproj]
  azure_core_storage_issue_249.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\Debug\azu
  re_core_storage_issue_249.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/core/azure-core/test/e2e/CMakeLists.txt
  azure_core_with_curl_bodyBuffer.cpp
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library [E:\GitHub\For
k\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\azure_core_with_curl.vcxproj]
  azure_core_with_curl.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\Debug\azure_core_
  with_curl.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/core/azure-core/test/e2e/CMakeLists.txt
  azure_core_with_curl_bodyStream.cpp
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library [E:\GitHub\For
k\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\azure_core_with_curl_stream.vcxproj]
  azure_core_with_curl_stream.vcxproj -> E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\Debug\azur
  e_core_with_curl_stream.exe
  Building Custom Rule E:/GitHub/Fork/azure-sdk-for-cpp/sdk/core/azure-core/test/e2e/CMakeLists.txt
  azure_hang.cpp
MSVCRTD.lib(initializers.obj) : warning LNK4098: defaultlib 'libcmtd.lib' conflicts with use of other libs; use /NODEFA
ULTLIB:library [E:\GitHub\Fork\azure-sdk-for-cpp\build\sdk\core\azure-core\test\e2e\azure_hang.vcxproj]
```

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #729

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,13 @@ option(BUILD_DOCUMENTATION "Create HTML based API documentation (requires Doxyge
 option(RUN_LONG_UNIT_TESTS "Tests that takes more than 5 minutes to complete. No effect if BUILD_TESTING is OFF" OFF)
 option(BUILD_STORAGE_SAMPLES "Build sample application for Azure Storage clients" OFF)
 
+# Use the static runtime libraries when building statically for consistency with vcpkg's
+# arch-windows-static triplets:
```
Comment (RickWinter) [inline CMakeLists.txt:22]:
```suggestion
# [arch]-windows-static triplets:
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +18,13 @@ option(BUILD_DOCUMENTATION "Create HTML based API documentation (requires Doxyge
...
+cmake_policy(SET CMP0091 NEW) # https://cmake.org/cmake/help/v3.15/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html
```
Comment (RickWinter) [inline CMakeLists.txt:23]: If this policy isn't available will CMAKE_MSVC_RUNTIME_LIBRARY be respected?  i.e. what happens if we are cmake 3.5 (ignore the fact other stuff in our cmake doesn't work on old cmake versions :) )

**Review by RickWinter — APPROVED** [review]

File: CMakeLists.txt [hunk 3]
```diff
@@ -18,6 +18,13 @@ option(BUILD_DOCUMENTATION "Create HTML based API documentation (requires Doxyge
...
+if (NOT DEFINED CMAKE_MSVC_RUNTIME_LIBRARY)
```
Comment (ahsonkhan) [inline CMakeLists.txt:24]: This was newly added in cmake 3.15 and our cmake currently requires 3.13 (with hopes to drop it lower). Using this is going in the other direction :p

I don't have a great alternative or solution, but maybe you can come up with something that works with an older version of cmake.

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
We probably want a way that works on 3.13 or lower.

**Comment by antkmsft** [comment]
@ahsonkhan , pre-cmake 3.15 fixes I've found do more harm than good in my opinion. They are ok for the application you know how you are building, but for the SDK library, which you may wish to build with multi-threaded static and dynamic CRT, they are too intrusive, unless we introduce a build flag called something like `MSVC_CRT = DYNAMIC | STATIC`.
pre-3.15 cmake did not handle this situation well - I think, it is better to show the warning. We used the same approach in C SDK.

File: CMakeLists.txt (hunk 2 above)
Comment (antkmsft) [reply CMakeLists.txt:23]: pre-cmake 3.15 fixes I've found do more harm than good in my opinion. They are ok for the application you know how you are building, but for the SDK library, which you may wish to build with multi-threaded static and dynamic CRT, they are too intrusive, unless we introduce a build flag called something like MSVC_CRT = DYNAMIC | STATIC.
pre-3.15 cmake did not handle this situation well - I think, it is better to show the warning. We used the same approach in C SDK.

**Comment by antkmsft** [comment]
This would be a good summary of possible solutions: https://stackoverflow.com/questions/14172856/compile-with-mt-instead-of-md-using-cmake
Note how all the pre-3.15 focus on overriding `/MD` with `/MT` - this is what I do not want us to do: you can chose to build with `/MT`, or `/MD`. Yes, you'll get warning when compiling with `/MD`, but at least we don't force you to a single option.

**Comment by ahsonkhan** [comment]
@barcharcraz, @BillyONeal - can you please suggest a good way to fix this LIBCMTD warning in C++, that works with cmake 3.15 or lower?

**Comment by vhvb1989** [comment]
Moving to Feb milestone since this still needs more investigation

**Comment by ahsonkhan** [comment]
@antkmsft this PR is over a month old. Are you still working on it, actively? If not, please close it for now and re-open once you are ready.

**Pull request closed without merging**